widgyts/dataset_viewer.py

```diff
@@ -0,0 +1,252 @@
+        return grid_views
+
+    @traitlets.default("r2_falloff")
+    def _r2_falloff_default(self):
+        x, y = np.mgrid[-0.5:0.5:32j, -0.5:0.5:32j]
+        r = (x**2 + y**2)**-0.5
+        r = np.clip(r, 0.0, 5.0)
+        r = (r - r.min())/(r.max() - r.min())
+        image_data = np.empty((32,32,4), dtype="f4")
+        image_data[:,:,:3] = r[:,:,None]
+        image_data[:,:,3] = 1.0
+        image_data = (image_data*255).astype("u1")
+        image_texture = pythreejs.BaseDataTexture(data=image_data)
+
+    @traitlets.default("colormap_texture")
+    def _colormap_texture_default(self):
+         import matplotlib.cm as mcm
+         viridis = mcm.get_cmap("viridis")
+         values = (viridis(np.mgrid[0.0:1.0:256j]) * 255).astype("u1")
+         values = np.stack([values[:,:],] * 256, axis=1).copy(order="C")
+         colormap_texture = pythreejs.BaseDataTexture(data = values)
+         return colormap_texture
+
+    @traitlets.default("renderer")
+    def _renderer_default(self):
+        center = tuple(self.ds.domain_center.in_units("code_length").d)
+        right = tuple(
+            (
+                self.ds.domain_right_edge
+                + (self.ds.domain_right_edge - self.ds.domain_center) * 2.0
+            )
+            .in_units("code_length")
+            .d
+        )
+        camera = pythreejs.PerspectiveCamera(
+            position=right, fov=20, children=[pythreejs.AmbientLight()]
+        )
+        scene = pythreejs.Scene(
+            children=[camera, pythreejs.AmbientLight(color="#dddddd")] + self.grid_views
+        )
+        orbit_control = pythreejs.OrbitControls(controlling=camera)
+        renderer = pythreejs.Renderer(
+            scene=scene,
+            camera=camera,
+            controls=[orbit_control],
+            width=400,
+            height=400,
+            background="black",
+            background_opacity=1,
+            antialias=True,
+        )
+        camera.lookAt(center)
+        orbit_control.target = center
+        renderer.layout.border = "1px solid darkgrey"
+        return renderer
+
+    def widget(self):
+        # Alright let's set this all up.
+        grid_contents = []
+        for i, view in enumerate(self.grid_views):
+            visible = ipywidgets.Checkbox(
+                value=view.visible, description="Level {}".format(i)
+            )
+            ipywidgets.jslink((visible, "value"), (view, "visible"))
+            color_picker = ipywidgets.ColorPicker(
+                value=view.material.color, concise=True
+            )
+            ipywidgets.jslink((color_picker, "value"), (view.material, "color"))
+            line_slider = ipywidgets.FloatSlider(
+                value=view.material.linewidth, min=0.0, max=10.0
+            )
+            ipywidgets.jslink((line_slider, "value"), (view.material, "linewidth"))
+            grid_contents.extend([visible, color_picker, line_slider])
+        return ipywidgets.HBox(
```

Comment:
It might be helpful to add an option that allows users to see a label of the view width as they zoom in and out, so they have a sense of dimension of each of the grids.

Reply:
So that's tricky to do, but I did explore it a bit.  The issue is that we're using a perspective camera, so in principle the view width is different at the near and far clipping planes.

Reply:
Longer term an annotation that sets up a line or something that can be moved would be possible to show distance

Reply:
I like that. We can open an issue for this so we address it at a later point.